cli/cli_app.go

```diff
@@ -274,16 +276,66 @@ func processHooks(hooks []config.Hook, terragruntOptions *options.TerragruntOpti
 	for _, curHook := range hooks {
 		allPreviousErrors := append(previousExecError, errorsOccurred...)
 		if shouldRunHook(curHook, terragruntOptions, allPreviousErrors...) {
+			var oldEnvMap map[string]string
+
 			terragruntOptions.Logger.Printf("Executing hook: %s", curHook.Name)
-			actionToExecute := curHook.Execute[0]
-			actionParams := curHook.Execute[1:]
-			possibleError := shell.RunShellCommand(terragruntOptions, actionToExecute, actionParams...)
+			if curHook.LoadEnvVars != nil {
+				terragruntOptions.Logger.Printf("Loading environment for hook: %s", curHook.Name)
+
+				actionToExecute := curHook.LoadEnvVars.Execute[0]
+				actionParams := curHook.LoadEnvVars.Execute[1:]
+				out, err := shell.RunShellCommandWithOutput(terragruntOptions, "", actionToExecute, actionParams...)
+				if err != nil {
+					terragruntOptions.Logger.Printf("Error running hook %s while loading environment variables with message: %s", curHook.Name, err.Error())
+					errorsOccurred = append(errorsOccurred, err)
+					continue
+				}
+				// parse out from script to map
+				m := make(map[string]string)
+				scanner := bufio.NewScanner(strings.NewReader(out.Stdout))
+				for scanner.Scan() {
+					txt := scanner.Text()
+					idx := strings.Index(txt, "=")
+					if idx > 0 {
+						key := strings.TrimSpace(txt[0:idx])
+						value := txt[idx+1 : len(txt)]
+						m[key] = value
+					} else {
+						terragruntOptions.Logger.Printf("Invalid key,value pair [%s]", txt)
+					}
+				}
+				targetEnvMap := terragruntOptions.Env
+				if curHook.LoadEnvVars.Transient {
+					oldEnvMap = terragruntOptions.Env
+					targetEnvMap := make(map[string]string)
+					for key, value := range terragruntOptions.Env {
+						targetEnvMap[key] = value
+					}
+					terragruntOptions.Env = targetEnvMap
+				}
+				if curHook.LoadEnvVars.Overwrite {
+					targetEnvMap = make(map[string]string)
+					terragruntOptions.Env = targetEnvMap
+				}
+				for key, value := range m {
+					targetEnvMap[key] = value
+				}
+			}
 
-			if possibleError != nil {
-				terragruntOptions.Logger.Printf("Error running hook %s with message: %s", curHook.Name, possibleError.Error())
-				errorsOccurred = append(errorsOccurred, possibleError)
+			if len(curHook.Execute) > 0 {
+				actionToExecute := curHook.Execute[0]
+				actionParams := curHook.Execute[1:]
+				possibleError := shell.RunShellCommand(terragruntOptions, actionToExecute, actionParams...)
+
+				if possibleError != nil {
+					terragruntOptions.Logger.Printf("Error running hook %s with message: %s", curHook.Name, possibleError.Error())
+					errorsOccurred = append(errorsOccurred, possibleError)
+				}
 			}
 
+			if oldEnvMap != nil {
+				terragruntOptions.Env = oldEnvMap
+			}
```

Comment:
Can you explain the thinking behind `oldEnvMap`?

Reply:
When you look at the explanation of the two flags `transient` and `overwrite` it's required to save the old environment variables and reset them on the `terragruntOptions` instance when `transient=true`

Reply:
Understood, thanks. As mentioned in comments above, instead of mutating `terragruntOptions`, I think `loadEnvVars` should return a clone of it with new env vars set as necessary. That changes the flow in this method to:

```go
terragruntOptionsWithEnvVars, err := loadEnvVars(terragruntOptions, curHook)
if err != nil {
  // ...
}
if len(curHook.Execute) > 0 {
  actionToExecute := curHook.Execute[0]
  actionParams := curHook.Execute[1:]
  possibleError := shell.RunShellCommand(terragruntOptionsWithEnvVars, actionToExecute, actionParams...)
  // ...
}
```